### ISSUE TITLE
Add nn.logmeanexp

### ISSUE THREAD

**Issue by carlosgmartin** [jax-ml/jax#30178]
Add to [`jax.nn`](https://docs.jax.dev/en/latest/jax.nn.html) a function `logmeanexp`, analogous to [`logsumexp`](https://docs.jax.dev/en/latest/_autosummary/jax.nn.logsumexp.html), that computes

$$ \text{logmeanexp}(x) = \log \frac{1}{n} \sum_{i=1}^n \exp x_i = \text{logsumexp}(x) - \log n $$

Some nontrivial code is needed to get the $\log n$ term right, as for [`mean`](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.mean.html). (For example, when `axis` is a tuple, or there is a `where` mask, or `keepdims` is used.)

Logsumexp is defined in other libraries:

- https://www.tensorflow.org/probability/api_docs/python/tfp/math/reduce_logmeanexp
- https://www.rdocumentation.org/packages/pomp/versions/6.3/topics/logmeanexp
- https://botorch.readthedocs.io/en/stable/utils.html#botorch.utils.safe_math.logmeanexp

It is also used to define the mellowmax operator:

$$ \text{mellowmax}_\omega(x) = \text{logmeanexp}(x \omega) / \omega $$

Here are some works that use MellowMax:

- [An alternative softmax operator for reinforcement learning (2016)](https://arxiv.org/abs/1612.05628)
- [Revisiting the softmax Bellman operator: New benefits and new perspective (2018)](https://arxiv.org/abs/1812.00456)
- [DeepMellow: removing the need for a target network in deep Q-learning (2019)](https://www.ijcai.org/proceedings/2019/379)
- [Adaptive temperature tuning for Mellowmax in deep reinforcement learning (2019)](https://cs.brown.edu/people/gdk/pubs/tuning_mellowmax_drlw.pdf)
- [Sample-efficient reinforcement learning with maximum entropy mellowmax episodic control (2019)](https://arxiv.org/abs/1911.09615)
- [Stabilizing Q learning via soft Mellowmax operator (2020)](https://arxiv.org/abs/2012.09456)
- [Damped Anderson mixing for deep reinforcement learning: Acceleration, convergence, and stabilization (2021)](https://proceedings.neurips.cc/paper_files/paper/2021/hash/1e79596878b2320cac26dd792a6c51c9-Abstract.html)
- [A weighted smooth Q-Learning algorithm (2025)](https://ieeexplore.ieee.org/abstract/document/10925426)

**Comment by jakevdp** [comment]
Sounds good - for what it's worth, the logic required for correct normalization / count is expressed here: https://github.com/jax-ml/jax/blob/6f8456e9fbf776dc85f5d7cf9b0bade9c658a76c/jax/_src/numpy/reductions.py#L888-L895
